extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,283 @@
+
+
+/** @const {Array} */
+//https://ctrlq.org/code/19899-google-translate-languages refer to google code
+const LOCALE_WORD = {
+  'zh-CN': ['年', '月', '天', '小时', '分钟', '秒'],
+  'zh-TW': ['年', '月', '天', '小時', '分鐘', '秒'],
+  'de': ['Jahren', 'Monaten', 'Tagen', 'Stunden', 'Minuten', 'Sekunden'],
+  'en': ['Years', 'Months', 'Days', 'Hours', 'Minutes', 'Seconds'],
+  'es': ['años', 'meses', 'días', 'horas', 'minutos', 'segundos'],
+  'fr': ['ans', 'mois', 'jours', 'heures', 'minutes', 'secondes'],
+  'id': ['tahun', 'bulan', 'hari', 'jam', 'menit', 'detik'],
+  'it': ['anni', 'mesi', 'giorni', 'ore', 'minuti', 'secondi'],
+  'ja': ['年', 'ヶ月', '日', '時間', '分', '秒'],
+  'ko': ['년', '달', '일', '시간', '분', '초'],
+  'nl': ['jaar', 'maanden', 'dagen', 'uur', 'minuten', 'seconden'],
+  'pt': ['anos', 'meses', 'dias', 'horas', 'minutos', 'segundos'],
+  'ru': ['год', 'месяц', 'день', 'час', 'минута', 'секунда'],
+  'th': ['ปี', 'เดือน', 'วัน', 'ชั่วโมง', 'นาที', 'วินาที'],
+  'tr': ['yıl', 'ay', 'gün', 'saat', 'dakika', 'saniye'],
+  'vi': ['năm', 'tháng', 'ngày', 'giờ', 'phút', 'giây'],
+};
+
+export class AmpDateCountdown extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const {function(!Element)} */
+    this.boundRendered_ = this.rendered_.bind(this);
+
+    //Note: One of end-date, timestamp-ms, timestamp-seconds is required.
+    /** @private {string} */
+    this.endDate_ = this.element.getAttribute('end-date');
+
+    /** @private {number} */
+    this.timestampMs_ = Number(this.element.getAttribute('timestamp-ms'));
+
+    /** @private {number} */
+    this.timestampSeconds_ = Number(this.element.getAttribute('timestamp-seconds'));
+
+    /** @private {number} */
+    this.offsetSeconds_ = Number(this.element.getAttribute('offset-seconds')) || DEFAULT_OFFSET_SECONDS;
+
+    /** @private {string} */
+    this.locale_ = (this.element.getAttribute('locale')) || DEFAULT_LOCALE;
+
+    /** @private {string} */
+    this.whenEnded_ = (this.element.getAttribute('when-ended')) || DEFAULT_WHEN_ENDED;
+
+    /** @private {string} */
+    this.biggestUnit_ = (this.element.getAttribute('biggest-unit')) || DEFAULT_BIGGEST_UNIT;
+
+    /** @private {!Object} */
+    this.localeWordList_ = this.getLocaleWord_(this.locale_);
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    const epoch = this.getEpoch_() + (this.offsetSeconds_ * 1000);
+    this.tickCountDown_(new Date(epoch) - new Date());
+  }
+
+  /** @override */
+  layoutCallback() {
+    const epoch = this.getEpoch_() + (this.offsetSeconds_ * 1000);
+    let differentBetween = new Date(epoch) - new Date() - 1000; //substract 1000 here because of buildCallback show the initial time
+    const delay = 1000;
+    return Promise.resolve(this.countDownTimer_ = this.win.setInterval(() => {
```

Comment:
Excellent, thank you for refactoring this. My last suggestions here are:

1. You'll need to initialize `this.countDownTimer_` in the constructor to make the linter happy
```js
/** @private {?number} */
this.countDownTimer_ = null;
```

2. You don't need to return anything in the Promise. It can just be `return Promise.resolve();`